haskell/asterius/defs.bzl

```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
+        ),
+    ])
+
+    ctx.actions.run_shell(
+        inputs = [ctx.file.dep, entrypoint_file, webpack_config],
+        outputs = [output_tar_file],
+        command = command,
+        env = {"PATH": ":".join(posix_toolchain.paths + [node_toolfiles[0].dirname])},
+        arguments = [
+            asterius_toolchain.ahc_dist.path,
+            ctx.file.dep.path,
+            file_copy_path,
+        ],
+        tools = asterius_toolchain.tools + node_toolfiles,  # + ctx.files.tools,
+    )
+
+    return [
+        DefaultInfo(
+            files = depset([output_tar_file]),
+            data_runfiles = ctx.attr.dep[0][DefaultInfo].data_runfiles,
+            default_runfiles = ctx.attr.dep[0][DefaultInfo].default_runfiles,
+        ),
+        AhcDistInfo(
+            target = ctx.attr.target,
+            webpack_config = webpack_config,
+            files = all_output_files,
+        ),
+    ]
+
+ahc_dist = rule(
+    _ahc_dist_impl,
+    attrs = {
+        "dep": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+            cfg = asterius_transition,
+            doc = """
+            The label of a haskell_binary, haskell_test or haskell_cabal_binary target.
+            """,
+        ),
+        "target": attr.string(
+            mandatory = True,
+            values = ["browser", "node"],
+            doc = """
+            Whether the build output is intended to run with node, or on the browser.
+            The asterius_webpack rule only accepts dependencies built for the browser.
+            """,
+        ),
+        "main_output": attr.string(
+            mandatory = True,
+            doc = "The name for the output file corresponding to the entrypoint. It must terminate by '.mjs'",
+        ),
+        "options": attr.string_list(
+            doc = "Other options to pass to ahc-dist",
+        ),
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
+        "_template": attr.label(
+            default = Label(_TEMPLATE),
+            allow_single_file = True,
+        ),
+    },
+    toolchains = [
+        "@rules_sh//sh/posix:toolchain_type",
+        "@rules_haskell//haskell/asterius:toolchain_type",
+        "@build_bazel_rules_nodejs//toolchains/node:toolchain_type",
+    ],
+    doc = "This rule transforms a haskell binary target into an archive containing javascript files.",
+)
+
+# extract files from the archive producted by ahc_dist in a subfolder,
+# to prevent conflicts as the library files have the same name.
+def _asterius_untar_impl(ctx):
+    posix_toolchain = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
+    subfolder_name = ctx.attr.subfolder_name
+    ahc_dist_info = ctx.attr.dep[AhcDistInfo]
+
+    output_files = [
+        ctx.actions.declare_file(paths.join(subfolder_name, name))
+        for name in ahc_dist_info.files
+    ]
```

Comment:
This is because the generated library files all have the same names, so there were conflicts if two `ahc_dist` rules were in the same directory. But maybe there is a better way around it, like generating the files in a subfolder, and then use `pkg_tar`?

Reply:
I see, thanks for explaining. Yes, I think a subfolder and emitting the files is a good solution to this. It avoids unecessary `tar/untar` operations. It also avoids leaving undeclared outputs in the execroot under `{workdir}`. These are not an issue for sandboxed builds, but with unsandboxed builds, like on Windows, these can cause issues.